drivers/SPI.h

```diff
@@ -74,9 +77,14 @@ namespace mbed {
  * @ingroup drivers
  */
 class SPI : private NonCopyable<SPI> {
-
+protected:
+    struct spi_peripheral_s {
+        SPIName name;
+        spi_t spi;
+        PlatformMutex *mutex;
```

Comment:
That's probably pushing too much stuff into the static allocation though. After more thought, this seems like a bad hybrid. There is a benefit to being fully-static, but if you're going to start doing some dynamic allocation, it may as well be fully dynamic to minimise overhead. I think rather than this static array, it would be better just to have the head of a linked list of peripherals, each of which with an embedded `PlatformMutex` and `spi_t`, so one heap block per peripheral.

Reply:
Ideally I would want the PlatformMutex to be statically allocated but this is causing issues with the constructor not being invoked when instantiated from a static array..